proposals/NativeInt.md

```diff
@@ -0,0 +1,263 @@
+Overrides and implementations may differ by `native int` and underlying type alone.
+Methods hide other methods that differ by `native int` and underlying type alone.
+
+### Miscellaneous
+
+`native int` types cannot be used as enum underlying types.
+```C#
+enum E : nint // error
+{
+}
+```
+
+`default(nint)` and `new nint()` are equivalent to `(nint)0`.
+
+`typeof(nint)` is `typeof(IntPtr)`.
+
+`sizeof(nint)` is supported and does not require compiling in an unsafe context. The value is not a compile-time constant. `sizeof(nint)` is implemented as `sizeof(IntPtr)` rather than `IntPtr.Size`.
+
+Compiler diagnostics for type references involving `nint` or `nuint` report `nint` or `nuint` rather than `IntPtr` or `UIntPtr`.
+
+### Metadata
+
+`nint` and `nuint` are represented in metadata as `System.IntPtr` and `System.UIntPtr`.
+
+Type references that include `nint` or `nuint` are emitted with a `System.Runtime.CompilerServices.NativeTypeAttribute` to indicate which parts of the type reference are native ints.
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    public sealed class NativeTypeAttribute : System.Attribute
+    {
+        public NativeTypeAttribute() { }
+        public NativeTypeAttribute(byte[] flags) { }
+    }
+}
+```
+
+The optional attribute argument contains a bit for each primitive type in the type reference.
+If there is a single primitive type, the parameter-less constructor can be used.
+
+```C#
+nuint A;                    // [NativeType] UIntPtr A
+(Stream, nint) B;           // [NativeType] ValueType<Stream, IntPtr> B
+Dictionary<object, nint> C; // [NativeType(new[] { 0, 1 })] Dictionary<object, IntPtr> C
+```
+
+_There are now four attributes that encode state for types within a type reference as an array of values: `DynamicAttribute`, `NullableAttribute`, `TupleElementNamesAttribute`, and `NativeTypeAttribute`. Each of these attributes uses a distinct mapping to generate the array elements._
+
+## Alternatives
+[alternatives]: #alternatives
+
+An alternative to the "type erasure" approach above is to introduce new types: `System.NativeInt` and `System.NativeUInt`.
+```C#
+public readonly struct NativeInt
+{
+    public IntPtr Value;
+}
+```
+Distinct types would allow overloading distinct from `IntPtr` and would allow distinct parsing and `ToString()`.
+But there would be more work for the CLR to handle these types efficiently which defeats the primary purpose of the feature - efficiency.
+And interop with existing native int code that uses `IntPtr` would be more difficult. 
+
+Another alternative is to add more native int support for `IntPtr` in the framework but without any specific compiler support.
+Any new conversions and arithmetic operations would be supported by the compiler automatically.
+But the language would not provide keywords, constants, or `checked` operations.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+What parts of the design are still undecided?
```

Comment:
>What parts of the design are still undecided [](start = 0, length = 44)

What is the common type between a value of type `nint` and a value of type `IntPtr`?  The common type between `G<nint>` and `G<IntPtr>`?  When both are bounds in type inference, which wins?